### PR TITLE
[Py3][ENG-491] Ultimate backwards compatible Python 3

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Make all possible backward compatible Python 3 changes.

## Changes
- minor dependency updates for addons.
- list wrap all dictviews
- stringify exception messages `e.message` -> `str(e)`
- change cmp argument to key for `sorted` 
- change asserts statement for new comparisons 
- remove unnecessary auth statements that were causing environmental errors
- remove unsed decorator `http_error_if_disk_saving_mode`
- changes ExternalProvider metaclass syntax to be both py2/3 compatible 

## QA Notes
- check order for filtering
- check that renaming publicly wiki
- Elastic search
- ownclould/twofactor have version bump

## Documentation

Not user facing

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-491

### CODE REVIEW

File: api/nodes/serializers.py [hunk 1]
```diff
@@ -1094,7 +1094,7 @@ def validate_data(self, node, user_id=None, full_name=None, email=None, index=No
             raise exceptions.ValidationError(detail='A user ID or full name must be provided to add a contributor.')
         if user_id and email:
             raise exceptions.ValidationError(detail='Do not provide an email when providing this user_id.')
-        if index > len(node.contributors):
+        if index is not None and index > len(node.contributors):
```
Comment (Johnetordoff) [inline api/nodes/serializers.py]: This is because of new py3 comparison rules.

File: website/identifiers/metadata.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# -*- coding: utf-8 -*-
```
Comment (brianjgeiger) [inline website/identifiers/metadata.py]: What's this file's story?

File: website/profile/views.py [hunk 3]
```diff
@@ -44,22 +42,14 @@
 logger = logging.getLogger(__name__)
 
 
-def date_or_none(date):
```
Comment (brianjgeiger) [inline website/profile/views.py]: Ah, so you removed the test because you removed this code, because it was unused?

File: website/project/decorators.py [hunk 4]
```diff
@@ -438,20 +438,6 @@ def wrapped(*args, **kwargs):
     # Return decorated function
     return wrapped
 
-def http_error_if_disk_saving_mode(func):
```
Comment (brianjgeiger) [inline website/project/decorators.py]: Seems like this would be worth mentioning in `Changes` as well as the other places there's deleted code.

File: api/caching/tests/test_caching.py [hunk 5]
```diff
@@ -171,7 +170,6 @@ def validate_keys(self, data, embed_keys):
             embed_keys = list()
 
         if 'errors' in data.keys():
-            print json.dumps(data, indent=4)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/caching/tests/test_caching.py]: No () I just removed it because it looked rogue.

File: api/institutions/authentication.py [hunk 6]
```diff
@@ -61,7 +61,7 @@ def authenticate(self, request):
                 options={'verify_exp': False},
                 algorithm='HS256',
             )
-        except (jwt.InvalidTokenError, TypeError):
+        except (jwt.InvalidTokenError, TypeError, jwe.exceptions.MalformedData):
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/institutions/authentication.py]: Yeah it's a byte/string conversion issue.

File: tasks/__init__.py [hunk 7]
```diff
@@ -801,7 +801,7 @@ def webpack(ctx, clean=False, watch=False, dev=False, colors=False):
 def build_js_config_files(ctx):
     from website import settings
     print('Building JS config files...')
-    with open(os.path.join(settings.STATIC_FOLDER, 'built', 'nodeCategories.json'), 'wb') as fp:
+    with open(os.path.join(settings.STATIC_FOLDER, 'built', 'nodeCategories.json'), 'w') as fp:
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tasks/__init__.py]: yes it's a byte string conversion issue.

File: tests/test_campaigns.py [hunk 8]
```diff
@@ -184,43 +184,44 @@ def setUp(self):
...
-            assert_equal(resp.status_code, http.OK)
+            assert_equal(resp.status_code, http_status.HTTP_200_OK)
 
 
 # tests for registration through campaigns
 class TestRegistrationThroughCampaigns(OsfTestCase):
 
     def setUp(self):
         super(TestRegistrationThroughCampaigns, self).setUp()
+        campaigns.get_campaigns()  # Set up global CAMPAIGNS
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_campaigns.py]: This is relative import issue CAMPAIGNS is a global.

File: tests/test_cas_authentication.py [hunk 9]
```diff
@@ -179,7 +179,7 @@ def test_application_token_revocation_succeeds(self):
             responses.Response(
                 responses.POST,
                 url,
-                body={'client_id': client_id,
+                json={'client_id': client_id,
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_cas_authentication.py]: Response library allows this for 2.7, but not 3

File: tests/test_conferences.py [hunk 10]
```diff
@@ -142,7 +141,7 @@ def setUp(self):
         self.conference = ConferenceFactory()
         self.body = 'dragon on my back'
         self.content = 'dragon attack'
-        self.attachment = StringIO(self.content)
+        self.attachment = BytesIO(self.content)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_conferences.py]: It's another byte string conversion issue the attachment is expected to be bytes in py3.

File: tests/test_events.py [hunk 11]
```diff
@@ -622,9 +622,9 @@ def setUp(self):
             'none': []
         }
         self.diff_1_3 = {email_transactional: ['d1234'], 'none': ['g1234'], email_digest: ['a1234']}
-        self.union_1_2 = {'email_transactional': ['e1234', 'd1234', 'k1234', 'f1234'],
+        self.union_1_2 = {'email_transactional': ['e1234', 'f1234', 'd1234', 'k1234'],
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_events.py]: Py3 dicts always retain order of dictionaries in terms of insertion, because of this the order has changed.

File: website/project/decorators.py (hunk 4 above)
Comment (Johnetordoff) [reply website/project/decorators.py]: This wasn't being used, I'll add it to changes.

File: addons/gitlab/tests/test_views.py [hunk 12]
```diff
@@ -195,14 +195,25 @@ def test_get_refs_sha_no_branch(self):
 
     # Tests for _check_permissions
     # make a user with no authorization; make sure check_permissions returns false
-    def test_permissions_no_auth(self):
+    @mock.patch('addons.gitlab.api.GitLabClient.repo')
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply addons/gitlab/tests/test_views.py]: Yes the autospec mocking didn't hold up under py3 new-style classes, IIRC.

File: api/nodes/serializers.py [hunk 13]
```diff
@@ -1227,7 +1227,7 @@ def validate_data(self, node, user_id=None, full_name=None, email=None, index=No
...
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/nodes/serializers.py]: Python 2: None > 1 == False
Python 3: None > 1 == TypeError

File: osf_tests/utils.py [hunk 14]
```diff
@@ -139,11 +139,10 @@ def mock_archive(project, schema=None, auth=None, data=None, parent=None,
         root_job.done = True
         root_job.save()
         sanction = registration.sanction
-        with contextlib.nested(
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf_tests/utils.py]: Yes, this was failing in Py3 and doesn't seem to be necessary anymore.

File: api/base/renderers.py [hunk 15]
```diff
@@ -9,9 +8,7 @@ class JSONRendererWithESISupport(JSONRenderer):
     def render(self, data, accepted_media_type=None, renderer_context=None):
         #  TODO: There should be a way to do this that is conditional on esi being requested and
         #  TODO: In such a way that it doesn't use regex unless there's absolutely no other way.
-        initial_rendering = super(JSONRendererWithESISupport, self).render(data, accepted_media_type, renderer_context)
-        augmented_rendering = re.sub(r'"<esi:include src=\\"(.*?)\\"\/>"', r'<esi:include src="\1"/>', initial_rendering)
-        return augmented_rendering
+        return super(JSONRendererWithESISupport, self).render(data, accepted_media_type, renderer_context)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api/base/renderers.py]: Initially I just removed, but I was able to revert and only change it by defining the charset this is something Django requires for py3 apparently.

File: tests/test_node_licenses.py [hunk 16]
```diff
@@ -103,6 +103,7 @@ def test_ensure_licenses_some_missing(self):
         assert_is_not_none(found)
 
     def test_ensure_licenses_updates_existing(self):
+        print(builtins)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply tests/test_node_licenses.py]: no :(

File: website/app.py [hunk 17]
```diff
@@ -95,16 +94,13 @@ def init_app(settings_module='website.settings', set_backends=True, routes=True,
     if app.config.get('IS_INITIALIZED', False) is True:
         return app
 
-    logger.info('Initializing the application from process {}, thread {}.'.format(
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply website/app.py]: This treading library doesn't work this way in Py3, since this was just a logging statement I decided to remove it.

File: website/identifiers/metadata.py (hunk 2 above)
Comment (Johnetordoff) [reply website/identifiers/metadata.py]: This just wasn't removed with other EZid code.

File: api_tests/nodes/views/test_node_detail.py [hunk 18]
```diff
@@ -1324,13 +1325,16 @@ def test_multiple_patch_requests_with_same_category_generates_one_log(
 
     def test_public_project_with_publicly_editable_wiki_turns_private(
             self, app, user, project_public, url_public, make_node_payload):
+        flask_context = Flask(__name__).test_request_context  # needed for pushing status message
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/nodes/views/test_node_detail.py]: This was apparently just unnecessary

File: api_tests/nodes/views/test_node_contributors_list.py [hunk 19]
```diff
@@ -154,7 +154,7 @@ def test_return(
             url_public, url_private, make_contrib_id):
 
         #   test_return_public_contributor_list_logged_in
-        res = app.get(url_public, auth=user_two.auth)
+        res = app.get(url_public)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/nodes/views/test_node_contributors_list.py]: This is the tricky part of this PR and I've reverted this and come up with an alternate fix. Problematically the auth is actually being sent via environmental variable `HTTP_AUTHORIZATION` which isn't being set in Py3. Fix for [TestApp](https://github.com/CenterForOpenScience/osf.io/pull/9047/commits/27a66835a065368165c3b6097194523b24f2a8ea#diff-b8a23469dba3532d9e9157217b0b46d3R74-R80) and [JSONAPITestApp](https://github.com/CenterForOpenScience/osf.io/pull/9047/commits/27a66835a065368165c3b6097194523b24f2a8ea#diff-44b2005fe01184a1089900dedd126fd4R117).

File: requirements.txt [hunk 20]
```diff
@@ -6,7 +6,7 @@
 # To install release requirements: inv requirements --release
 
 invoke==0.15.0
-Werkzeug==0.10.4  # pyup: ignore # newer versions break local development
+Werkzeug==0.12.1  # pyup: ignore # newer versions break local development
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply requirements.txt]: This can be reverted because of the `auth` arg changes so back to `Werkzeug==0.10.4` for now.

File: osf/models/node.py [hunk 21]
```diff
@@ -2472,6 +2472,16 @@ def is_bookmark_collection(self):
         """For v1 compat"""
         return False
 
+    class Meta:
+        # custom permissions for use in the OSF Admin App
+        permissions = (
+            ('view_node', 'Can view node details'),
+        )
+
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/node.py]: this was just a copy paste error, so I'm removing it.

File: osf/models/node.py [hunk 22]
```diff
@@ -2472,6 +2472,16 @@ def is_bookmark_collection(self):
         """For v1 compat"""
         return False
 
+    class Meta:
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/node.py]: this was just a copy paste error, so I'm removing it.

File: api_tests/preprints/views/test_preprint_contributors_list.py [hunk 23]
```diff
@@ -128,7 +128,7 @@ def test_return(
             url_published, url_unpublished, make_contrib_id):
 
         #   test_return_published_contributor_list_logged_in
-        res = app.get(url_published, auth=user_two.auth)
+        res = app.get(url_published)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/preprints/views/test_preprint_contributors_list.py]: [explaination](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: api_tests/search/views/test_views.py [hunk 24]
```diff
@@ -262,7 +262,7 @@ def test_search_components(
         assert component.title in res
 
         # test_search_public_component_auth
-        res = app.get(url_component_search, auth=user)
+        res = app.get(url_component_search)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/search/views/test_views.py]: [explanation](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: api_tests/search/views/test_views.py [hunk 25]
```diff
@@ -272,7 +272,7 @@ def test_search_components(
         assert component.title in res
 
         # test_search_public_component_contributor
-        res = app.get(url_component_search, auth=user_two)
+        res = app.get(url_component_search)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/search/views/test_views.py]: [explanation](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: api_tests/search/views/test_views.py [hunk 26]
```diff
@@ -287,12 +287,12 @@ def test_search_components(
...
 
         # test_search_private_component_auth
-        res = app.get(url_component_search, auth=user)
+        res = app.get(url_component_search)
         assert res.status_code == 200
         assert component_private.title not in res
 
         # test_search_private_component_contributor
-        res = app.get(url_component_search, auth=user_two)
+        res = app.get(url_component_search)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/search/views/test_views.py]: [explanation](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: api_tests/search/views/test_views.py [hunk 27]
```diff
@@ -374,7 +374,7 @@ def test_search_files(
         assert file_component.name in res
 
         # test_search_public_file_auth
-        res = app.get(url_file_search, auth=user)
+        res = app.get(url_file_search)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/search/views/test_views.py]: [explanation](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: api_tests/search/views/test_views.py [hunk 28]
```diff
@@ -384,7 +384,7 @@ def test_search_files(
         assert file_component.name in res
 
         # test_search_public_file_contributor
-        res = app.get(url_file_search, auth=user_one)
+        res = app.get(url_file_search)
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply api_tests/search/views/test_views.py]: [explanation](https://github.com/CenterForOpenScience/osf.io/pull/9047#discussion_r326790354)

File: osf/migrations/0136_preprint_node_divorce.py [hunk 29]
```diff
@@ -91,7 +90,6 @@ def format_group(self, name):
 
 def emit_signals(state, schema):
     # this is to make sure that the permissions created earlier exist!
-    emit_post_migrate_signal(2, False, 'default')
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/migrations/0136_preprint_node_divorce.py]: I reverted this and it had no effect/

**Comment by brianjgeiger** [comment]
@Johnetordoff While I'm reviewing, can you double check to ensure that anything I had a question as to whether it was meant to be part of this PR and is still in is reflected in the `Changes` section of the PR? If you're confident, that's cool, I just want to make sure it's documented.

File: api_tests/nodes/views/test_node_detail.py [hunk 30]
```diff
@@ -1329,8 +1329,9 @@ def test_public_project_with_publicly_editable_wiki_turns_private(
         res = app.patch_json_api(
             url_public,
             make_node_payload(project_public, {'public': False}),
-            auth=user.auth  # self.user is creator/admin
+            auth=user.auth  # self.user is creator/admind
```
Comment (brianjgeiger) [inline api_tests/nodes/views/test_node_detail.py]: `admind`?

File: requirements.txt [hunk 31]
```diff
@@ -21,7 +21,7 @@ git+https://github.com/cos-forks/celery@v4.1.1+cos0
 kombu==4.2.0
 amqp==2.2.2
 vine==1.1.4
-itsdangerous==0.24
+itsdangerous==1.1.0
```
Comment (brianjgeiger) [inline requirements.txt]: Should this be listed in `Changes`?

Comment (Johnetordoff) [reply requirements.txt]: Apparently not necesary anyomore so removed.

**Comment by Johnetordoff** [comment]
@brianjgeiger ?

**Comment by brianjgeiger** [comment]
@Johnetordoff Yes?

**Comment by Johnetordoff** [comment]
If this is Final Review now, who is remaining to review it?

**Comment by brianjgeiger** [comment]
The ticket is closed as done. I figured it would just end up being me, although @mfraezz may decide he really wants to look through it if that's his thing.

**Comment by Johnetordoff** [comment]
Is this code going to merged though? Are we going to wait until the next sprint? It is difficult to keep this from conflicting.

**Comment by brianjgeiger** [comment]
It'll get merged when Rabia is ready for it. We can prioritize it, though, during the next sprint planning. There's not a lot planned to be merged for the rest of this sprint, so hopefully there won't be too many updates to do.

**Comment by brianjgeiger** [comment]
@Johnetordoff I'm afraid there's a merge conflict

File: api/caching/tests/test_caching.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mfraezz) [reply api/caching/tests/test_caching.py]: Aside: I think this whole file might be rogue. It's not in `api_tests` but `api/[...]/tests`, and we didn't get varnish running

File: website/project/views/drafts.py [hunk 32]
```diff
@@ -39,7 +39,7 @@ def get_schema_or_fail(schema_name, schema_version):
     try:
         meta_schema = RegistrationSchema.objects.get(name=schema_name, schema_version=schema_version)
     except RegistrationSchema.DoesNotExist:
-        raise HTTPError(http_status.HTTP_200_OK, data=dict(
```
Comment (mfraezz) [inline website/project/views/drafts.py]: 😐

File: osf/migrations/0137_transfer_preprint_service_permissions.py [hunk 33]
```diff
@@ -3,12 +3,9 @@
 from __future__ import unicode_literals
 
 from django.db import migrations
-from django.core.management.sql import emit_post_migrate_signal
```
Comment (mfraezz) [inline osf/migrations/0137_transfer_preprint_service_permissions.py]: Why was this removed? Is it no longer supported / done automatically between migrations in Django2 / Py3?
